### PR TITLE
Fix CA2256 false positive for nested types

### LINKED ISSUE #5964: CA2256 false positive (?) on nested interface
### Analyzer

**Diagnostic ID**: [CA2256](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2256): `All members declared in parent interfaces must have an implementation in a DynamicInterfaceCastableImplementation-attributed interface`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 6.0.201](https://dotnet.microsoft.com/download/dotnet/6.0)

### Describe the bug

CA2256 is (I believe) firing incorrectly on nested interfaces attributed as `[DynamicInterfaceCastableImplementation]`.

### Steps To Reproduce

```csharp
public interface ICLRRuntimeLocator
{
    [DynamicInterfaceCastableImplementation]
    internal interface ICLRRuntimeLocatorWrapper : ICLRRuntimeLocator
    {
        ulong ICLRRuntimeLocator.GetRuntimeBase()
        {
            return 0;
        }
    }

    public static readonly Guid IID_ICLRRuntimeLocator = new("b760bf44-9377-4597-8be7-58083bdc5146");

    ulong GetRuntimeBase();
}
```

### Expected behavior

No warning.

### Actual behavior

Warning fires on `ICLRRuntimeLocatorWrapper`.

## Additional context

Moving `ICLRRuntimeLocatorWrapper` out of `ICLRRuntimeLocator` so it is no longer nested gets rid of the warning.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #5964

### CODE REVIEW

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/DynamicInterfaceCastableImplementationTests.cs [hunk 1]
```diff
@@ -1132,6 +1133,53 @@ static void Method2(I2 @this, int i)
...
+        }
+    }
+
+    void M();
+}
+";
+            await VerifyCSAnalyzerAsync(source);
+        }
+
+        [Fact(Skip = "TODO: How should the codefix behave.")]
```
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/InteropServices/DynamicInterfaceCastableImplementationTests.cs:1162]: **This needs to be fixed before merging the PR.**

**Review by Youssef1313 — COMMENTED** [review]
@jkoritzinsky @mavasani This is ready for review.

**Comment by Youssef1313** [comment]
There is a bug here, which is unfortunately on Roslyn side. Added a failing test and will see if I can workaround it.

**Comment by Youssef1313** [comment]
Note to self when I get back to this. https://github.com/dotnet/roslyn/pull/61134 might affect https://github.com/dotnet/roslyn-analyzers/issues/5964#issuecomment-1096899383 ?

**Review by jkoritzinsky — APPROVED** [review]

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/InteropServices/CSharpDynamicInterfaceCastableImplementation.Fixer.cs [hunk 2]
```diff
@@ -73,7 +71,7 @@ public sealed class CSharpDynamicInterfaceCastableImplementationFixer : DynamicI
                     return null;
                 }
                 var methodDecl = generator.MethodDeclaration(method);
```
Comment (Evangelink) [inline src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/InteropServices/CSharpDynamicInterfaceCastableImplementation.Fixer.cs:73]: Could we rename this as `methodDeclaration`? I am not fan of abbreviations and it's not used consistently here (I know it was there before but there is enough changes to ask for it :)).

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DynamicInterfaceCastableImplementation.Fixer.cs [hunk 3]
```diff
@@ -41,7 +41,7 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
                     context.RegisterCodeFix(
                         CodeAction.Create(
                             MicrosoftNetCoreAnalyzersResources.ImplementInterfacesOnDynamicCastableImplementation,
-                            async ct => await ImplementInterfacesOnDynamicCastableImplementationAsync(declaration, context.Document, context.CancellationToken).ConfigureAwait(false),
+                            async ct => await ImplementInterfacesOnDynamicCastableImplementationAsync(root, declaration, context.Document, generator, ct).ConfigureAwait(false),
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DynamicInterfaceCastableImplementation.Fixer.cs:44]: Most of the other codefixes are using the lambda token so I think that's ok. Thanks for pointing it out.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DynamicInterfaceCastableImplementation.cs [hunk 4]
```diff
@@ -99,7 +99,8 @@ private static void AnalyzeType(SymbolAnalysisContext context, INamedTypeSymbol
             {
                 foreach (var member in iface.GetMembers())
                 {
-                    if (!member.IsStatic
+                    if (member.IsAbstract &&
+                        member.Kind != SymbolKind.NamedType
```
Comment (Evangelink) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DynamicInterfaceCastableImplementation.cs:103]: Nit: to match the other operator placement

```suggestion
                    if (member.IsAbstract
                        && member.Kind != SymbolKind.NamedType
```

**Review by Evangelink — APPROVED** [review]
LGTM. Thanks @Youssef1313

**Comment by Evangelink** [comment]
@Youssef1313 I have rerun jobs but maybe you will have to merge main to avoid the CI issue.

**Comment by Youssef1313** [comment]
I merged latest main now. If it failed again (which I feel will happen), then this is probably the result of a bug fix in Roslyn and the failing tests need to be updated.

File: src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/InteropServices/CSharpDynamicInterfaceCastableImplementation.Fixer.cs (hunk 2 above)
Comment (Youssef1313) [reply src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/InteropServices/CSharpDynamicInterfaceCastableImplementation.Fixer.cs:73]: Sure. Will do

**Comment by Evangelink** [comment]
The logs are mentioning a confusion around MEF.

**Comment by Youssef1313** [comment]
@Evangelink I was seeing other failures:

image

**Comment by Evangelink** [comment]
Failing again :s

**Comment by Youssef1313** [comment]
@Evangelink

https://github.com/dotnet/roslyn-analyzers/blob/3cd7f620e70b3f2643ca513637fa62f0ad6930fb/src/Roslyn.Diagnostics.Analyzers/UnitTests/ExportedPartsShouldHaveImportingConstructorTests.cs#L37

It looks like `SystemCompositionReference` is no longer needed, and the type can be found without it? I'm not sure how upgrading `Microsoft.CodeAnalysis` causes that.

**Comment by Evangelink** [comment]
Hum I suspect this is linked to `The type 'ExportAttribute' exists in both 'System.Composition.AttributedModel, Version=1.0.31.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' and 'System.Composition.AttributedModel, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'` where one dll is found in v1 and the other in v6. But I am not quite sure what's happening.

Also I have noticed that some tests are working locally and failing on CI because of some missing `ReferenceAssemblies`. I don't remember we had different behaviour locally or remotely before.

**Comment by Evangelink** [comment]
> For the tests I run locally, they have same behavior in CI. Can you point to a specific test where you see different behavior?

Yes for example: [`d7fb0dd` (#5874)](https://github.com/dotnet/roslyn-analyzers/pull/5874/commits/d7fb0dd98c8edc31fa15b14b9f06ab5fef831f09). I had to fix the reference while the test was working well without the change for me locally.

> Somehow we are broken by [dotnet/roslyn@2e2b033](https://github.com/dotnet/roslyn/commit/2e2b033f7764efd6b25b7bb674df662cdb5b1e65)

Thanks, so we now need to find what is bringing the v1.

**Comment by Youssef1313** [comment]
> Thanks, so we now need to find what is bringing the v1.

It was the explicit reference assembly in the test, which I removed now. For me the question is the other way around, how the 6.0.0 got into the test if the Microsoft.CodeAnalysis reference assembly for the test is explicitly set to 3.0.0, but the answer isn't really important.

We should have a green CI now.

**Comment by Youssef1313** [comment]
> Yes for example: [`d7fb0dd` (#5874)](https://github.com/dotnet/roslyn-analyzers/pull/5874/commits/d7fb0dd98c8edc31fa15b14b9f06ab5fef831f09). I had to fix the reference while the test was working well without the change for me locally.
> 
> >

Test project is a multi-target-framework (both net472 and netcoreapp3.1). In `AdditionalMetadataReferences` (`CreateDefaultReferenceAssemblies` specifically) we have `#if` directives based on the TFM.

You might have run the test locally on one TFM, and the one failing in CI was the other.
I'm not aware of anyway to run tests locally for both TFMs other than running test from command-line. VS don't let me choose the TFM for testing, even if I choose it here:

image

I'm not sure if that's what you were encountering, but it's a possibility.

**Pull request merged**